internal/cryptotest/aead.go

```diff
@@ -0,0 +1,387 @@
+// the key must be fixed.
+type MakeAEAD func() (cipher.AEAD, error)
+
+// TestAEAD performs a set of tests on cipher.AEAD implementations, checking
+// the documented requirements of NonceSize, Overhead, Seal and Open.
+func TestAEAD(t *testing.T, mAEAD MakeAEAD) {
+	aead, err := mAEAD()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("Roundtrip", func(t *testing.T) {
+
+		// Test all combinations of plaintext and additional data lengths.
+		for _, ptLen := range lengths {
+			for _, adLen := range lengths {
+				t.Run(fmt.Sprintf("Plaintext-Length=%d,AddData-Length=%d", ptLen, adLen), func(t *testing.T) {
+					rng := newRandReader(t)
+
+					nonce := make([]byte, aead.NonceSize())
+					rng.Read(nonce)
+
+					before, addData := make([]byte, adLen), make([]byte, ptLen)
+					rng.Read(before)
+					rng.Read(addData)
+
+					ciphertext := sealMsg(t, aead, nil, nonce, before, addData)
+					after := openWithoutError(t, aead, nil, nonce, ciphertext, addData)
+
+					if !bytes.Equal(after, before) {
+						t.Errorf("plaintext is different after a seal/open cycle; got %s, want %s", truncateHex(after), truncateHex(before))
+					}
+				})
+			}
+		}
+	})
+
+	t.Run("InputNotModified", func(t *testing.T) {
+
+		// Test all combinations of plaintext and additional data lengths.
+		for _, ptLen := range lengths {
+			for _, adLen := range lengths {
+				t.Run(fmt.Sprintf("Plaintext-Length=%d,AddData-Length=%d", ptLen, adLen), func(t *testing.T) {
+					t.Run("Seal", func(t *testing.T) {
+						rng := newRandReader(t)
+
+						nonce := make([]byte, aead.NonceSize())
+						rng.Read(nonce)
+
+						src, before := make([]byte, ptLen), make([]byte, ptLen)
+						rng.Read(src)
+						copy(before, src)
+
+						addData := make([]byte, adLen)
+						rng.Read(addData)
+
+						sealMsg(t, aead, nil, nonce, src, addData)
+						if !bytes.Equal(src, before) {
+							t.Errorf("Seal modified src; got %s, want %s", truncateHex(src), truncateHex(before))
+						}
+					})
+
+					t.Run("Open", func(t *testing.T) {
+						rng := newRandReader(t)
+
+						nonce := make([]byte, aead.NonceSize())
+						rng.Read(nonce)
+
+						plaintext, addData := make([]byte, ptLen), make([]byte, adLen)
+						rng.Read(plaintext)
+						rng.Read(addData)
+
+						// Record the ciphertext that shouldn't be modified as the input of
+						// Open.
+						ciphertext := sealMsg(t, aead, nil, nonce, plaintext, addData)
+						before := make([]byte, len(ciphertext))
+						copy(before, ciphertext)
+
+						openWithoutError(t, aead, nil, nonce, ciphertext, addData)
+						if !bytes.Equal(ciphertext, before) {
+							t.Errorf("Open modified src; got %s, want %s", truncateHex(ciphertext), truncateHex(before))
+						}
+					})
+				})
+			}
+		}
+	})
+
+	t.Run("BufferOverlap", func(t *testing.T) {
+
+		// Test all combinations of plaintext and additional data lengths.
+		for _, ptLen := range lengths {
+			if ptLen <= 1 { // We need enough room for an overlap to occur.
```

Comment:
Nit, but this is the first time it's mentioned in the test case so relatively important. 😄 (1 would be enough for an overlap?)

```suggestion
			if ptLen <= 1 { // We need enough room for an inexact overlap to occur.
```